Mabofoul/step by step

NoneSince opened the pull request:
the 5th commit (c78612cc87d8f89a7738c947f9061759e9b6d363) is a revert to start again from your master. I did so to push again the changes I have done but make each step visible by giving it its own commit. now it is easier for you to check file comparisons.
regarding the utility function split() that I added, I didnt test it, but I ran the whole program on the examples and it worked.

NoneSince commented:
Conversation regarding this PR is under PR #3

Thomas-de-Bock reviewed (commented):
Please try compiling and assembling the construct code before committing. In "strchr" findchr is macrod to sil because we cannot compare a single byte with a 64 bit register, which is why findchr is macrod to the 8 bit region of the register. This is also why when assembling it will give `strchr.asm:10: error: mismatch in operand sizes`

NoneSince commented:
Sorry, I did not think of that! I will revert this change.

Also, about the "split takes alot of lines" issue, in a next PR im moving it down:
#include ...
...
static void to_lower(string& str);

class IsAnyOf {
private:
  string chars;
public:
...
};

template <typename Predicate>
static void split(vector<string>& result, const string& input, const Predicate& pred,
                  const bool& compress_adj_delims=true, const bool& strip=true);
...
implemenation of deconstruct functions...
...
// ----- ----- ----- ----- ----- ----- helper functions impl ----- ----- ----- ----- -----
implementation for split and stuff

Thomas-de-Bock commented:
> Sorry, I did not think of that! I will revert this change.
> 
> Also, about the "split takes alot of lines" issue, in a next PR im moving it down: #include ... ... static void to_lower(string& str);
> 
> class IsAnyOf { private: string chars; public: ... };
> 
> template static void split(vector& result, const string& input, const Predicate& pred, const bool& compress_adj_delims=true, const bool& strip=true); ... implemenation of deconstruct functions... ... // ----- ----- ----- ----- ----- ----- helper functions impl ----- ----- ----- ----- ----- implementation for split and stuff

I think as long ONLY the required string functions are added and they're placed neatly so as not to cause confusion it might be okay to just fit it into the deconstruct file. I would like it if you did not make them template functions however. I'll take a look at the next PR.

Thomas-de-Bock commented:
To get all my notes straight, if:
- The required string functions are implemented neat and simple, and placed so it is not confusing.
- The construct code is reverted
- The changes to the reg_to_str function are removed (to me it's the same as hardcoding a string into a string split function, because it's only used once with a set string. Sure maybe it won't be used anywhere else, but I want to avoid global variables in functions as much as we possible can)
I will merge the PR.

NoneSince commented:
also, i have just noted reg_to_str() is used only in reconstruct.cpp, so it should be static.
What is the code style you drive for? I for real suggest you use "my" style, because that's literally how everyone i have seen in person writes code, and the only devs with your style I saw online were JSdevelopers. In JS there are arrow functions and many wrappers so you might want to collapse the brackets. And here, functions and classes are few so you pay no penalty for giving theie scope a special look.

Thomas-de-Bock commented:
reg_to_str being static makes sense yeah since it will likely only ever be used to reconstruct code so I like that yes. My main issue with making functions that are only used in their respective source file static is that then we would have to do it for all of them, but if you are up to that I don't mind. Also I am aware that this isn't the usual style, I do not have this habit from js haha. I just think having a curly brace on its own seperate line makes sense when closing a function, as it's the only indicator (besides indentation) that a function has ended. But for opening a function you already have the function decleration as a signal that a new function has been opened. It's just preference but I see no reason to change it for everything.

Thomas-de-Bock commented:
Also  I really like how you did the string functions here. Instead of cluttering the header file which is supposed to be a nice overview of functionality, you just quickly declared them at the start of the cpp file. I'll look over it properly tommorow, but this is good.

Thomas-de-Bock commented:
Though why wouldn't you just make the string split function return a pointer to the result?

NoneSince commented:
> Though why wouldn't you just make the string split function return a pointer to the result?

I did that because I was mimicking boost::split, and I didnt think of doing it as you say.
I will define the vector inside the function, and return a copy.
note that there is no need for pointers, vector  inside the function will be copied outside and then deleted, so you can return a vector instead of vector*.
if you meant the function receives a vector and returns a pointer to it, then that's meaningless. simply get a reference and write inside it (exactly like now)

about code style, i will keep up with your style, since you started all of that. If I were the owner, I would run my style (also, indentation is 4 spaces), so now I cede this matter to you, with full acceptance in-sha-allah.

NoneSince commented:
I see a problem in parsing commands with memory addresses, like "mov eax, dword ptr [ebx]".
I split a line by spaces and comma, then take first 3 words. I will try to fix that by splitting by comma first, and all what is to right will be put together, and the words to left will be the command itself and arg1.
And an extra check is to see if there are brackets to denote memory address, and if both arg1 and arg2 are memory, its a syntax error (in assembly, not that we wanna prevent for convenience)

NoneSince commented:
another thing: why do you do to_lower() to every line you parse? it makes strlwr.con example store the string already in lower case, which proves the action is wrong.
strlwr.con: teststring db "HeLlO WoRlD", 0
strlwr.asm: teststring db "hello world", 0

also, there is no colon after "teststring". isn't that wrong?

Thomas-de-Bock commented:
Pulled and compiled the branch, running construct on the example strlwr.con gives a SIGSEGV (Address boundary error). And the name of the compiled executable is construct.exe, it should be dependant on the OS. (Linux executables don't have an extension.)

NoneSince commented:
regarding construct.exe, windows must have the .exe even if you dont write any extention, while linux isn't affected with the extension, so i did put the extention. you can roll with it and fix that later if you know (it is not a "fix" since there is no problem, you just prefer the name under linux be "construct") .
about the error, I don't know, but since them i have fixed some stuff and bugs that I added. i think i will do a PR that only changes style (no logic) for you to quickly test and push, because I have alot of free time so I tamper with the code and grow further each day

NoneSince commented:
now everything works: i ran your original master on the 3 examples, and ran my PR on these examples, and compared the output files (didnt run them, and need to). the files are the same, which means that my code is a step forward.
regarding ".exe" extention, keep it and merge the PR, because this reaaaally drives us back to freeze all the PR work for a further notice. as i said, im building over it

NoneSince commented:
@Thomas-de-Bock please see the last update

Thomas-de-Bock commented:
I'll take a look at it tomorrow. Also for the exe extension thing, that's not up for debate. It only had no file extension before because the build was never meant for windows. The exe extension is exclusive to windows executables, this is not my preference.

NoneSince commented:
"I'll merge it if that's fixed"... why not merge it then remove the .exe?
also as I said, I don't run the output asm file, I only compare it to what was produced before, so even if the out asm is broken, my PR is still fine and ready for merge, since I achieved the current state in a better way.
